### PR TITLE
[RELAY][OP]Strided slice

### PR DESCRIPTION
**Pull request by siju-samuel**
#1799

Thanks for contributing to TVM!   Please refer to guideline https://docs.tvm.ai/contribute/ for useful information and tips. After the pull request is submitted, please request code reviews from others in the community.

### CODE REVIEW

**Comment by siju-samuel** [comment]
@MarisaKirisame @yzhliu @yuruofeifei @srkreddy1238 @tqchen please review.

File: include/tvm/relay/attrs/transform.h [hunk 1]
```diff
@@ -82,6 +82,22 @@ struct FullAttrs : public tvm::AttrsNode<FullAttrs> {
...
+struct StridedSliceAttrs : public tvm::AttrsNode<StridedSliceAttrs> {
+  Array<IndexExpr> begin;
+  Array<IndexExpr> end;
+  Array<IndexExpr> stride;
+
+  TVM_DECLARE_ATTRS(StridedSliceAttrs, "relay.attrs.StridedSliceAttrs") {
+    TVM_ATTR_FIELD(begin)
+        .describe("Indices for begin of slice");
+    TVM_ATTR_FIELD(end)
+        .describe("Indices for end of the slice");
```
Comment (MarisaKirisame) [inline include/tvm/relay/attrs/transform.h]: describe inclusive, exclusive?

File: src/relay/op/tensor/transform.cc [hunk 2]
```diff
@@ -578,5 +578,123 @@ Examples::
...
+
+      CHECK(reporter->Assert(stride_vec[i] < 0) ?
+              reporter->Assert(end < begin) : reporter->Assert(begin < end))
+        << ": Input [Begin=" << begin_vec[i] << ", End=" << end_vec[i]
+        << "] is invalid for axis=" << i;
+      oshape[i] = slice_size;
+  }
+
+  reporter->Assign(types[1], TensorTypeNode::make(oshape, data->dtype));
+  return true;
```
Comment (MarisaKirisame) [inline src/relay/op/tensor/transform.cc]: undef min/max somewhere

**Comment by tqchen** [comment]
please rebase against master after https://github.com/dmlc/tvm/pull/1934 to make use of the newly introduced API and add test-case to make sure text format works. Thanks!

**Comment by siju-samuel** [comment]
@tqchen this can be merged? Anything else need to be done? can you please review once again and let me know. Thanks.

**Comment by tqchen** [comment]
@MarisaKirisame please https://docs.tvm.ai/contribute/code_review.html#approve-and-request-changes-explicitly

File: src/relay/op/tensor/transform.cc [hunk 3]
```diff
@@ -891,6 +891,120 @@ RELAY_REGISTER_OP("broadcast_to_like")
...
+  for (auto i : param->stride) {
+    stride_vec.push_back(i);
+  }
+  for (auto i = stride_vec.size(); i < num_axis; ++i) {
+    stride_vec.push_back(1);
+  }
+  std::vector<IndexExpr> oshape(dshape.size());
+
+  for (size_t i = 0; i < num_axis; ++i) {
+      auto begin_range = reporter->Assert(stride_vec[i] < 0) ? -1 : 0;
```
Comment (tqchen) [inline src/relay/op/tensor/transform.cc]: using Assert here is not reliable, as if stride_vec is symbolic, then assert does not reflect anything.

File: include/tvm/relay/attrs/transform.h [hunk 4]
```diff
@@ -122,6 +122,21 @@ struct SplitAttrs : public tvm::AttrsNode<SplitAttrs> {
   }
 };
 
+/*! \brief Attributes for StridedSlice operator */
+struct StridedSliceAttrs : public tvm::AttrsNode<StridedSliceAttrs> {
+  Array<IndexExpr> begin;
```
Comment (tqchen) [inline include/tvm/relay/attrs/transform.h]: Let us change ```Array<IndexExpr> -> Array<Integer>``` for now, as integer is really what we can do reliably so far

File: include/tvm/relay/attrs/transform.h [hunk 5]
```diff
@@ -122,6 +122,21 @@ struct SplitAttrs : public tvm::AttrsNode<SplitAttrs> {
...
+  Array<IndexExpr> end;
+  Array<IndexExpr> stride;
```
Comment (tqchen) [inline include/tvm/relay/attrs/transform.h]: stride->strides as per https://www.tensorflow.org/api_docs/python/tf/strided_slice

File: src/relay/op/tensor/transform.cc [hunk 6]
```diff
@@ -891,6 +891,120 @@ RELAY_REGISTER_OP("broadcast_to_like")
...
+      auto end_range = reporter->Assert(stride_vec[i] < 0) ? dshape[i] - 1 : dshape[i];
+      auto begin = reporter->Assert(begin_vec[i] < 0) ? dshape[i] + begin_vec[i] : begin_vec[i];
+      auto end = reporter->Assert(end_vec[i] < 0) ? dshape[i] + end_vec[i] : end_vec[i];
+
+      begin = min(max(begin, begin_range), end_range);
```
Comment (tqchen) [inline src/relay/op/tensor/transform.cc]: use std::min std::max,

File: tests/python/relay/test_op_level4.py [hunk 7]
```diff
@@ -90,10 +90,31 @@ def test_reduce_functions():
         verify_reduce(func, (128, 24, 128), (0, 2), False, False, (24,))
         verify_reduce(func, (128, 24, 128), (0, 1), True, False, (1, 1, 128))
         verify_reduce(func, (128, 24, 128), (0, 2), True, False, (1, 24, 1))
+def verify_strided_slice(data, begin, end, stride, output):
```
Comment (tqchen) [inline tests/python/relay/test_op_level4.py]: newline between functions

**Comment by tqchen** [comment]
Given that there is still some gap and we need this OP in quickly, I am opening a followup https://github.com/dmlc/tvm/pull/2094 which is based on this PR.

**Comment by tqchen** [comment]
Thanks @siju-samuel @MarisaKirisame

**Pull request closed without merging**